File: src/heroscribe/list/Read.java
/*
  HeroScribe
  Copyright (C) 2002-2004 Flavio Chierichetti and Valerio Chierichetti
   
  This program is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License version 2 (not
  later versions) as published by the Free Software Foundation.
 
  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.
 
  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

package org.lightless.heroscribe.list;

import java.io.File;

import org.xml.sax.*;
import org.xml.sax.helpers.DefaultHandler;
import javax.xml.parsers.*;

public class Read extends DefaultHandler {
	private List objects;
	
	private boolean onBoard;
	private LObject piece;
	private LBoard board;
	private StringBuffer content;

	public Read(File file) throws Exception {
		super();

		objects = new List();

		SAXParserFactory factory = SAXParserFactory.newInstance();
		factory.setValidating(true);

		SAXParser saxParser = factory.newSAXParser();
		saxParser.parse(file, this);
	}

	public List getObjects() {
		return objects;
	}

	/* --- */

	public InputSource resolveEntity(String publicId, String systemId) {
		if ( publicId.equals("-//org.lightless//HeroScribe Object List 1.5//EN") )
			return new InputSource("DtdXsd/objectList-1.5.dtd");
		else
			return null;
	}

	public void error(SAXParseException e) throws SAXException {
		throw new SAXException(e);
	}

	public void startDocument() {
		content = new StringBuffer();
		onBoard = false;
	}

	public void startElement(String uri, String localName, String qName, Attributes attrs) throws SAXException {
		content = new StringBuffer();

		if (qName == "objectList") {
			objects.version = attrs.getValue("version");
			
			if ( ! objects.version.equals( org.lightless.heroscribe.Constants.version ) )
				throw new SAXException("HeroScribe's and Objects.xml's version numbers don't match.");
			
			objects.vectorPrefix = attrs.getValue("vectorPrefix");
			objects.vectorSuffix = attrs.getValue("vectorSuffix");

			objects.rasterPrefix = attrs.getValue("rasterPrefix");
			objects.rasterSuffix = attrs.getValue("rasterSuffix");

			objects.samplePrefix = attrs.getValue("samplePrefix");
			objects.sampleSuffix = attrs.getValue("sampleSuffix");
		} else if (qName == "kind") {
			objects.kinds.add(
				new Kind(attrs.getValue("id"), attrs.getValue("name")) );
		} else if (qName == "board") {
			board = new LBoard(Integer.parseInt(attrs.getValue("width")),
				Integer.parseInt(attrs.getValue("height")));
			
			board.borderDoorsOffset =
				Float.parseFloat(attrs.getValue("borderDoorsOffset"));
			
			board.adjacentBoardsOffset =
				Float.parseFloat(attrs.getValue("adjacentBoardsOffset"));
			
			onBoard = true;
		} else if (qName == "object") {
			piece = new LObject();

			piece.id = attrs.getValue("id");
			piece.name = attrs.getValue("name");

			piece.kind = attrs.getValue("kind");
			
			piece.door = Boolean.valueOf( attrs.getValue("door") ).booleanValue();
			piece.trap = Boolean.valueOf( attrs.getValue("trap") ).booleanValue();
			
			piece.width = Integer.parseInt(attrs.getValue("width"));
			piece.height = Integer.parseInt(attrs.getValue("height"));
			
			piece.zorder = Float.parseFloat(attrs.getValue("zorder"));
			
			piece.note = null;
		} else if (qName == "icon") {
			Icon icon = new Icon();
			
			icon.path = attrs.getValue("path");
			icon.xoffset = Float.parseFloat( attrs.getValue("xoffset") );
			icon.yoffset = Float.parseFloat( attrs.getValue("yoffset") );
			
			icon.original = Boolean.valueOf( attrs.getValue("original") ).booleanValue();
			
			if ( onBoard )
				board.putIcon(icon, attrs.getValue("region"));
			else
				piece.putIcon(icon, attrs.getValue("region"));
		} else if (qName == "corridor") {
			if ( onBoard ) {
				int width, height;
				int left, top;

				width = Integer.parseInt(attrs.getValue("width"));
				height = Integer.parseInt(attrs.getValue("height"));
				left = Integer.parseInt(attrs.getValue("left"));
				top = Integer.parseInt(attrs.getValue("top"));

				if ( left + width - 1 > board.width || left < 1 ||
					top + height - 1 > board.height || top < 1 )
					throw new SAXException("Corridors: out of border");

				for (int i = 0; i < width; i++)
					for (int j = 0; j < height; j++)
						board.corridors[i + left][j + top] = true;
				
			}
		}
	}

	public void characters(char[] ch, int start, int length) {
		content.append(ch, start, length);
	}

	public void endElement(String uri, String localName, String qName) throws SAXException {
		if (qName == "board" ) {
			if ( !board.region.containsKey("Europe") ||
				 !board.region.containsKey("USA") )
				 throw new SAXException("There should be both icons for each board.");

			objects.board = board;
				
			onBoard = false;
		} else if ( qName =="object") {
			if ( !piece.region.containsKey("Europe") ||
				 !piece.region.containsKey("USA") )
				 throw new SAXException("There should be both icons for each object.");
			
			objects.list.put(piece.id, piece);
		} else if (qName == "note") {
			piece.note = new String(content);
		}
	}

	public void endDocument() {
		content = null;
		piece = null;
	}
}

File: src/heroscribe/list/Read.py
# -*- coding: utf-8 -*-
"""
Created on Sat Mar 23 16:02:50 2019

@author: Andreas
"""

from xml.dom.minidom import parse
import xml.dom.minidom
from pathlib import Path

from src.heroscribe.list.LObject import LObject
from src.heroscribe.list.List import  List
from src.heroscribe.list.LBoard import LBoard
from src.heroscribe.list.Kind import Kind
from src.heroscribe.list.Icon import Icon

from src.heroscribe import Constants

class Read():

    def __init__(self, file_loc):

        self.object_list = List()
        self.piece = LObject()

        # There will be a board of type LBoard later, but not right now.
        #self.board
        self.content = ''
        # gives the state during parsing if we are on a board or not.
        self.on_board = False


        DomTree = xml.dom.minidom.parse(str(file_loc))
        object_xml = DomTree.documentElement
        #for key in object_xml._attrs.keys():
        self.object_list.version = object_xml.getAttribute('version')
        self.object_list.vector_prefix = object_xml.getAttribute('vectorPrefix')
        self.object_list.vector_suffix = object_xml.getAttribute('vectorSuffix')
        self.object_list.raster_prefix = object_xml.getAttribute('rasterPrefix')
        self.object_list.raster_suffix = object_xml.getAttribute('rasterSuffix')
        self.object_list.sample_prefix = object_xml.getAttribute('samplePrefix')
        self.object_list.sample_suffix = object_xml.getAttribute('sampleSuffix')

        # heroscribe folders
        for kind in object_xml.getElementsByTagName("kind"):
            kind_id = kind.getAttribute('id')
            kind_name = kind.getAttribute('name')
            self.object_list.kinds.add(Kind(kind_id, kind_name))


        # game boards
        for xmlboard in object_xml.getElementsByTagName("board"):
            self.object_list.board = self.read_xml_board(xmlboard)

        # other game pieces
        for xmlpiece in object_xml.getElementsByTagName("object"):
            piece = self.get_game_piece(xmlpiece)
            self.object_list.object_list[piece.id] = piece





    def get_game_piece(self, xmlpiece):
        piece = LObject()
        piece.id = xmlpiece.getAttribute("id")
        piece.name = xmlpiece.getAttribute("name")
        piece.kind = xmlpiece.getAttribute("kind")

        piece.door = xmlpiece.getAttribute("door")
        piece.trap = xmlpiece.getAttribute("trap")
        piece.u_trap = xmlpiece.getAttribute("untraceabletrap")

        piece.width = xmlpiece.getAttribute("width")
        piece.height = xmlpiece.getAttribute("height")
        piece.zorder = xmlpiece.getAttribute("zorder")
        piece.mini_icon = xmlpiece.getAttribute("miniIcon")

        # TODO: put tag dictionary into the xml file and read them here,
        # for better filter functions
        piece.tags  = xmlpiece.getAttribute("tags")

        for xmlicon in xmlpiece.getElementsByTagName("icon"):
            icon, region = self.read_xml_icon(xmlicon)
            piece.put_icon(icon, region)

        piece.note = self.get_notes(xmlpiece)
        return piece

    def get_notes(self, xml_anything):
        ''' extracts all text from all notes nodes in xml_anything'''
        notes = ''
        if xml_anything.getElementsByTagName("note"):
            for xmlnote in xml_anything.getElementsByTagName("note"):
                notes = notes + xmlnote.childNodes[0]._data
        return notes



    def read_xml_icon(self, xmlicon):
        '''reads an icon into the icon data structure and returns it
        '''

        path = xmlicon.getAttribute('path')
        xoffset = xmlicon.getAttribute('xoffset')
        if bool(xoffset):
            xoffset = float(xoffset)
        else:
            xoffset = 0
        yoffset = xmlicon.getAttribute('yoffset')
        if bool(yoffset):
            yoffset = float(yoffset)
        else:
            yoffset = 0
        original = xmlicon.getAttribute('original')
        original = original.lower in ['true']
        region = xmlicon.getAttribute('region')
        return Icon(path, xoffset, yoffset, original), region


    def read_xml_board(self, xmlboard):
        board = LBoard(xmlboard.getAttribute('width'),
                              xmlboard.getAttribute('height'))
        board.border_doors_offset = float(xmlboard.getAttribute("borderDoorsOffset"))
        board.adjacent_boards_offset = float(xmlboard.getAttribute('adjacentBoardsOffset'))
        board.name = xmlboard.getAttribute('name')
        board.default = xmlboard.getAttribute('default')

        for xmlicon in xmlboard.getElementsByTagName("icon"):
            icon, region = self.read_xml_icon(xmlicon)
            board.put_icon(icon, region)

        # corridors are defined to make them a little darker in the US version.
        for xmlcorridor in xmlboard.getElementsByTagName("corridor"):
            left = int(xmlcorridor.getAttribute("left"))
            top = int(xmlcorridor.getAttribute("top"))
            width = int(xmlcorridor.getAttribute("width"))
            height = int(xmlcorridor.getAttribute("height"))
            if (left + width -1 > board.width
            or left < 1
            or top + height -1 > board.width
            or top < 1):
                # TODO: Make a better understandable error message
                raise ValueError("Corridors: out of board border")

            # mark squares on board as corridor
            # TODO: Check corridor markings
            for x in range(0, width):
                for y in range(0, height):
                    board.corridors[x + left -1 ][y + top -1] = True


        for xmlroom in xmlboard.getElementsByTagName("room"):
            # TODO: For questimator extension, read rooms here
            pass
        return board



file_loc = Path('C:/Users/Andreas/25 Heroquest/Heroscribe 2/Sample.xml')
readfile = Read(file_loc)







